### PR TITLE
Version: 0.1.0

### PR DESCRIPTION
**Pull request by HW13**
### Summary ###

staging-auth is an [express](https://github.com/expressjs/express) middleware, that utilizes [express-session](https://github.com/expressjs/session) to provide user authentication in staging web apps via a login form.

### Test Plan ###

- Clone and then `cd` into the repo.
- Run `yarn` and then `yarn build`.
- Run `yarn link`.
- Create a new project directory and run:
```
$ yarn init -y
$ yarn link "staging-auth"
$ yarn add express
$ yarn add staging-auth
$ export STAGING_AUTH_USER="user"
$ export STAGING_AUTH_PASSWORD="password"
```
- In your new project create an `index.js` and add the follow code:
```
var express = require('express');
var stagingAuth = require('staging-auth').default;

var app = express();

app.use(stagingAuth('secret_string'));

app.get('/', function (req, res) {
   res.send('Hello World');
});

app.listen(3000, function (err) {
  if (err) {
    return console.error(err);
  }
  return console.info(`Server running on http://localhost:3000`);
});

```
- In your terminal, `cd` into your new project repo and run `node index.js`.
- Open a browser window and go to `http://localhost:3000` and try out the login.
   - The username is `user` and the password is `password`.

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+  cookie: {
+    expires: 600000,
+  },
+};
+
+const defaultTemplate = {
+  templateString: stringTemplate,
+};
+
+// TODO(HW13): Add JSdoc comments.
```
Comment (jkirkpatrick24) [inline src/index.js]: Lets add these comments in before we merge this. While its fresh in your mind.

File: package.json [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    "eslint": "^4.9.0",
+    "eslint-plugin-flowtype": "^2.39.1",
+    "eslint-plugin-import": "^2.14.0",
+    "eslint-plugin-mocha": "^5.2.0",
+    "flow-bin": "^0.81.0",
+    "mocha": "^5.2.0",
+    "rimraf": "^2.6.2",
+    "sinon": "^6.3.4"
+  },
+  "npmName": "staging-auth",
```
Comment (jkirkpatrick24) [inline package.json]: Just generally speaking, we are going to publish this under the autotelic org right? Should we namespace this? eg `autotelic/express-staging-authentication`

File: src/index.js [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+export const createStagingAuthRouter = (
+  router: express$Router = express.Router(),
```
Comment (jkirkpatrick24) [inline src/index.js]: This argument is the least likely to be overwritten so it should be last.

File: src/index.js [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+  getMiddleware: (config: SessionConfig) => express$Middleware[] = getDefaultMiddleware,
+  authMiddleware: AuthMiddleware = stagingAuthMiddleware,
+  routerGet: StagingAuthGetMethod = getStagingAuth,
+  routerPost: StagingAuthPostMethod = postStagingAuth,
+) => (
+  sessionSecret: string,
+  redirectPath: string = '/',
+  loginPath: string = '/staging-auth',
```
Comment (jkirkpatrick24) [inline src/index.js]: Should this be configurable?

File: src/index.js [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+// TODO(HW13): Add JSdoc comments.
+export const createStagingAuthRouter = (
+  router: express$Router = express.Router(),
+  getMiddleware: (config: SessionConfig) => express$Middleware[] = getDefaultMiddleware,
+  authMiddleware: AuthMiddleware = stagingAuthMiddleware,
+  routerGet: StagingAuthGetMethod = getStagingAuth,
+  routerPost: StagingAuthPostMethod = postStagingAuth,
+) => (
+  sessionSecret: string,
+  redirectPath: string = '/',
```
Comment (jkirkpatrick24) [inline src/index.js]: [nit] What if I want to navigate directly to `mysite.com/some-url`? Very much a nit though.

File: src/index.js [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+  ));
+
+  router.get(loginPath, routerGet(loginTemplate));
+
+  router.post(loginPath, routerPost(authPassword, authUser, loginPath, redirectPath));
+
+  return router;
+};
+
+export default createStagingAuthRouter();
```
Comment (jkirkpatrick24) [inline src/index.js]: This export pattern seems a bit funny to me? Why not just have the user call the function when they import it?

File: src/types.js [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+  resave: boolean,
+  saveUninitialized: boolean,
+  cookie: {
+    expires: number,
+  },
+};
+
+// Only one of the following is required.
+// If both are provided, then the templateString will be used.
+export type LoginTemplate = {
```
Comment (jkirkpatrick24) [inline src/types.js]: I wonder if there is a stricter way to express this with flow 🤔

File: test/getStagingAuth.test.js [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+      sendFile: sendFileSpy,
+    };
+  })
+
+  describe('If only a templateString is provided.', () => {
+    it('Should call .send with the templateString.', () => {
+      // Arrange
+      loginTemplate.templateString = templateString;
+      routerGet = getStagingAuth(loginTemplate);
+      // Act
```
Comment (jkirkpatrick24) [inline test/getStagingAuth.test.js]: [nit] There should be a newline between Arrange, Act, Assert blocks. Here and elsewhere.

File: test/index.test.js [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+  let sessionConfig;
+  let router
+  let routerGetSpy;
+  let routerPostSpy;
+  let getSpy;
+  let postSpy;
+  let useSpy;
+
+  // Arrange
+  beforeEach(() => {
```
Comment (jkirkpatrick24) [inline test/index.test.js]: I think this beforeEach block is a bit much. Are all of these variables are used in every test. I think if there are only used once its fine to have them declared within the test. Makes it easier to understand.

File: test/index.test.js [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+    };
+    sessionConfig = { secret: '123' };
+    stagingAuth = createStagingAuthRouter(
+      router,
+      getMiddlewareSpy,
+      authMiddlewareSpy,
+      routerGetSpy,
+      routerPostSpy
+    );
+    callStagingAuth = () => stagingAuth(
```
Comment (jkirkpatrick24) [inline test/index.test.js]: Imo I find wrapping this in a callback a bit odd. Do you feel strongly about it? I think it makes the test a bit harder to understand but maybe thats just me.

File: test/middleware.test.js [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+      session: {},
+      path: '',
+    };
+    res = {
+      clearCookie: clearCookieSpy,
+      redirect: redirectSpy,
+    };
+  })
+
+  describe('If req.path is the loginPath.', () => {
```
Comment (jkirkpatrick24) [inline test/middleware.test.js]: These tests could all be written as a table.

File: test/postStagingAuth.test.js [hunk 12]
```diff
@@ -0,0 +1,88 @@
+import { assert } from 'chai';
+import sinon from 'sinon';
+
+import postStagingAuth from '../src/routing/postStagingAuth';
+
+describe('Routing: postStagingAuth.', () => {
```
Comment (jkirkpatrick24) [inline test/postStagingAuth.test.js]: Should these files be colocated? That way we don't need the `Routing: <function>` which is a bit confusing.

File: test/postStagingAuth.test.js [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+      // Act
+      routerPost(req, res);
+      // Assert
+      assert.isOk(redirectSpy.calledOnceWith(loginPath));
+      assert.isOk(req.session.user !== authUser);
+      assert.isOk(statusSpy.calledOnceWith(401));
+    });
+  });
+
+  describe('authUser and authPassword do not match the req.body username and password.', () => {
```
Comment (jkirkpatrick24) [inline test/postStagingAuth.test.js]: Probably a good table test 😄

File: src/index.js (hunk 4 above)
Comment (HW13) [reply src/index.js]: It is configurable, it just defaults to `/staging-auth`.

File: src/index.js (hunk 5 above)
Comment (HW13) [reply src/index.js]: The `redirectPath` is passed directly to `res.redirect`, so you can use a path or url as the param. I'll remove path from the param name.

File: src/index.js (hunk 6 above)
Comment (HW13) [reply src/index.js]: For testing I wrap the function that creates the middleware in a function that injects the router, middleware, handlers, etc. I call the function when I export so the user doesn't have to call the imported function an additional time before they can pass it to `app.use`.

File: src/index.js (hunk 3 above)
Comment (HW13) [reply src/index.js]: None of those are meant to be overwritten, they are only passed as params for testing purposes. The function returned by this wrapper has the params that are intended to be configurable.

File: src/routing/handleGet.js [hunk 14]
```diff
@@ -0,0 +1,42 @@
+// @flow
+
+import type { LoginOptions } from '../types';
+
+/**
+ * Config for the staging-auth login page.
+ * (Either a templateString or templatePath should be provided not both.)
```
Comment (jkirkpatrick24) [inline src/routing/handleGet.js]: [nit] Punctuation

File: src/routing/handlePost.js [hunk 15]
```diff
@@ -2,7 +2,17 @@
 
 import get from 'lodash.get';
 
-export default (
+/**
+ * Function that an express middleware that handles post requests.
```
Comment (jkirkpatrick24) [inline src/routing/handlePost.js]: [nit] wut?

**Pull request merged**